src/array_string.rs

```diff
@@ -295,6 +295,36 @@ impl<const CAP: usize> ArrayString<CAP>
         Ok(())
     }
 
+    pub fn try_push_iterator<I: IntoIterator<Item = char>>(&mut self, value: I) -> Result<(), CapacityError<()>> {
+        let mut len = self.len();
+        let value = value.into_iter();
+        if value.size_hint().0 > self.capacity() - len {
+            return Err(CapacityError::new(()));
+        }
+
+        unsafe {
+            let mut ptr = self.as_mut_ptr().add(self.len());
+            for c in value {
+                let remaining_cap = self.capacity() - len;
+                match encode_utf8(c, ptr, remaining_cap) {
+                    Ok(n) => {
+                        len += n;
+                        ptr = ptr.add(n);
+                    }
+                    Err(_) => return Err(CapacityError::new(())),
+                }
+            }
+            self.set_len(len);
```

Comment:
Why is this implemented as `unsafe` instead of using `ArrayString::try_push`?

Reply:
Because I think that this function should be as performant as possible, and calling `try_push` will modify the `len` on every iteration. This way we can set it once when we're done, which seems to be more efficient.

Reply:
Could you write a quick benchmark verifying that this improves performance over the naive `try_push` loop?

Reply:
Sure, here are results on my machune:

```
test try_extend_bench ... bench:          28 ns/iter (+/- 2) = 146285 MB/s
test try_push_loop    ... bench:         336 ns/iter (+/- 13) = 12190 MB/s
```

I also fixed the other benches (at they were all optimized on my machine in just one LOAD of a capacity constant)